### PR TITLE
docs: Add extension categories to extension docs

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: docs: Add extension categories to extension docs
Additional Description:

Add information about known  extension categories (types) to extension documentation

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: api/envoy/config/listener/v3/listener_components.proto [hunk 1]
```diff
@@ -34,6 +34,7 @@ message Filter {
 
   // The name of the filter to instantiate. The name must match a
   // :ref:`supported filter <config_network_filters>`.
+  // [#protodoc-extension: Filter]
```
Comment (htuch) [inline api/envoy/config/listener/v3/listener_components.proto:37]: Network filter?

File: tools/protodoc/protodoc.py [hunk 2]
```diff
@@ -132,7 +132,10 @@ def FormatCommentWithAnnotations(comment, type_name=''):
   if annotations.EXTENSION_ANNOTATION in comment.annotations:
     extension = comment.annotations[annotations.EXTENSION_ANNOTATION]
     formatted_extension = FormatExtension(extension)
-  return annotations.WithoutAnnotations(StripLeadingSpace(comment.raw) + '\n') + formatted_extension
+  formatted_extension_type = ''
+  if annotations.EXTENSION_TYPE_ANNOTATION in comment.annotations:
+    formatted_extension_type = "INFO ABOUT THE EXTENSION TYPE 8)"
```
Comment (htuch) [inline tools/protodoc/protodoc.py:137]: I assume this becomes more elaborated when you are done the draft work :)

**Review by htuch — COMMENTED** [review]
Nice, this looks like the right direction.

File: tools/protodoc/protodoc.py (hunk 2 above)
Comment (phlax) [reply tools/protodoc/protodoc.py:137]: yep, hacking atm to see what is significant

File: api/envoy/config/listener/v3/listener_components.proto (hunk 1 above)
Comment (mattklein123) [reply api/envoy/config/listener/v3/listener_components.proto:37]: I think I would recommend using the extension category we use in the code for registration? E.g.

https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/include/envoy/server/filter_config.h#L128

@phlax this the category that is printed as part of this loop here:

https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/source/server/server.cc#L325-L328

Essentially I think we need to go from category -> name/type_url which we implicitly have in the registered factories structure but we will need to either get out of a C++ utility or more optimally encode in bazel somehow.

Comment (mattklein123) [reply api/envoy/config/listener/v3/listener_components.proto:37]: (I realize that the extension categories are Envoy specific, but IMO that is fine for now until we sort out how to do per-client documentation.)

Comment (phlax) [reply api/envoy/config/listener/v3/listener_components.proto:37]: so, trying to figure out the existing categories, because ultimately its these that need to match

im seeing that as part of protodoc an `extensions.db` json file is created, and checking the keys of this these do appear to be the categories that are marked in the protos

eg:

``` console
$ git grep "\[\#extension: envoy.access_loggers.http_grpc\]" api/
api/envoy/config/accesslog/v2/als.proto:// [#extension: envoy.access_loggers.http_grpc]
api/envoy/extensions/access_loggers/grpc/v3/als.proto:// [#extension: envoy.access_loggers.http_grpc]
api/envoy/extensions/access_loggers/grpc/v4alpha/als.proto:// [#extension: envoy.access_loggers.http_grpc]
```

what that doesnt give me afaict is the list of extensions for each extension

im wondering whether i can do this with some kind of (existing) bazel query/command to resolve these extensions, or whether as the above comment seems to suggest these need to be parsed from factories in the source code (or i guess a running binary)

Comment (phlax) [reply api/envoy/config/listener/v3/listener_components.proto:37]: @htucch any input on ^^ would be very helpful - i can start trying to hack protodoc to parse the extensions if this is required, but would like to confirm it is before trying to do so

Comment (htuch) [reply api/envoy/config/listener/v3/listener_components.proto:37]: The protos aren't marking categories, they are using the legacy extension names that we previously had prior to type URLs and `typed_config` becoming the standard way to indicate extension by type.

I think we should make a mandatory field in `envoy_cc_extension` that includes the extension category (which should match one of Envoy's already known types). This will allow `bazel query` analysis.

File: bazel/envoy_library.bzl [hunk 3]
```diff
@@ -80,6 +80,8 @@ EXTENSION_STATUS_VALUES = [
 def envoy_cc_extension(
         name,
         security_posture,
+        # Make this mandatory once all extensions have had their cat added.
+        category = None,
```
Comment (phlax) [inline bazel/envoy_library.bzl:84]: @htuch so iiuc we need this (altho made mandatory)

and...

File: source/extensions/filters/network/client_ssl_auth/BUILD [hunk 4]
```diff
@@ -16,6 +16,7 @@ envoy_cc_library(
     name = "client_ssl_auth",
     srcs = ["client_ssl_auth.cc"],
     hdrs = ["client_ssl_auth.h"],
+    category = "envoy.filters.network",
```
Comment (phlax) [inline source/extensions/filters/network/client_ssl_auth/BUILD:19]: ...this to mark categories in extension build files

File: bazel/envoy_library.bzl (hunk 3 above)
Comment (htuch) [reply bazel/envoy_library.bzl:84]: Yep

File: source/extensions/filters/network/client_ssl_auth/BUILD (hunk 4 above)
Comment (htuch) [reply source/extensions/filters/network/client_ssl_auth/BUILD:19]: Right. This will allow you to add the extension category to the metadata in the extension DB built by https://github.com/envoyproxy/envoy/blob/main/docs/generate_extension_db.py.

https://github.com/envoyproxy/envoy/blob/main/tools/protodoc/protodoc.py can be extended to indicate category on the extension metadata formatting.

If you want to build a list of extensions for a given category by Bazel path, I think that's easy enough to do with `bazel query` and the compulsory category field. It's a bit trickier to map from category to configuration proto or configuration RST. One idea might be to extend the type database, built with https://github.com/envoyproxy/envoy/blob/main/tools/type_whisperer/typedb_gen.py to also indicate the extension that a given proto type is responsible for configuring, but the downside is that this takes a very general type database and starts to special case it for extensions (I'm OK with this though).

There are other hackier possibilities. E.g. building all the docs and scraping with grep etc. the extension name from RST and figuring out which path this RST is at, building some RST mapping database.

File: docs/generate_extension_db.py [hunk 5]
```diff
@@ -49,6 +50,7 @@ def GetExtensionMetadata(target):
       'security_posture': security_posture,
       'undocumented': False if IsMissing(undocumented) else bool(undocumented),
       'status': 'stable' if IsMissing(status) else status,
+      'categories': 'nocategory' if IsMissing(categories) else categories,
```
Comment (phlax) [inline docs/generate_extension_db.py:53]: no need for `ismissing` as its mandatory

File: docs/generate_extension_cats_db.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    with open(extension_db) as f:
+        data = json.load(f)
+
+    categories = {}
+    for k, v in data.items():
+        for cat in v['categories']:
+            categories[cat] = categories.get(cat, [])
+            categories[cat].append(k)
+
+    print(categories)
```
Comment (phlax) [inline docs/generate_extension_cats_db.py:20]: remove

File: docs/generate_extension_db.py [hunk 7]
```diff
@@ -37,10 +37,11 @@ def GetExtensionMetadata(target):
   if not BUILDOZER_PATH:
     raise ExtensionDbError('Buildozer not found!')
   r = subprocess.run(
-      [BUILDOZER_PATH, '-stdout', 'print security_posture status undocumented', target],
+      [BUILDOZER_PATH, '-stdout', 'print security_posture status undocumented category', target],
       stdout=subprocess.PIPE,
       stderr=subprocess.PIPE)
-  security_posture, status, undocumented = r.stdout.decode('utf-8').strip().split(' ')
+  security_posture, status, undocumented = r.stdout.decode('utf-8').strip().split(' ')[:3]
```
Comment (phlax) [inline docs/generate_extension_db.py:43]: add `ismissing` fail for category

File: docs/build.sh [hunk 8]
```diff
@@ -75,6 +77,10 @@ BAZEL_BUILD_OPTIONS+=(
 # metadata, based on the envoy_cc_extension() Bazel target attributes.
 ./docs/generate_extension_db.py "${EXTENSION_DB_PATH}"
 
+# Generate extension category database.
+./docs/generate_extension_cats_db.py "${EXTENSION_DB_PATH}" "${EXTENSION_CAT_DB_PATH}"
```
Comment (phlax) [inline docs/build.sh:81]: rename to ..cat_db

File: tools/protodoc/protodoc.py [hunk 9]
```diff
@@ -204,6 +211,23 @@ def FormatExtension(extension):
     exit(1)  # Raising the error buries the above message in tracebacks.
 
 
+def FormatExtensionType(extension_type):
+  """Format extension metadata as RST.
+
+  Args:
+    extension: the name of the extension, e.g. com.acme.foo.
```
Comment (phlax) [inline tools/protodoc/protodoc.py:218]: `extension_type`

File: api/envoy/config/listener/v3/listener_components.proto (hunk 1 above)
Comment (phlax) [reply api/envoy/config/listener/v3/listener_components.proto]: this part i think is resolved by #14744

File: source/extensions/filters/network/client_ssl_auth/BUILD (hunk 4 above)
Comment (phlax) [reply source/extensions/filters/network/client_ssl_auth/BUILD]: atm i have used the generated extension db to generate an extension type db that can then be used in protodoc

this is a bit docs specific, so i might need to rethink to accomodate generating (vscode) jsonschemas but for now this should give us the abillity to back/forward ref extensions <> types

**Comment by phlax** [comment]
@htuch - a ~quick summary of the status of this PR...

### matching extensions <> categories

ive added the annotation for the extension points where i believed best - some of these may be wrong.

i added "markers" in protodoc so we can easily find the extensions/categories.

The categories - ie extension points - can be seen here:

https://storage.googleapis.com/envoy-pr/327f548/docs/search.html?q=EXTENSION_TYPE_MARKER&check_keywords=yes&area=default#

this link should allow you to see where the extension points are, and click through to the actual extensions (and back again)

i have a feeling some of the extensions may also be incorrectly categorized also - i will happily update this PR with fixes if that is the case.

### rendering extensions/categories and links

so far i have done the minimum required to render the links, and not given too much thought to either how they are rendered or code quality.

i will tidy this up over the next few days

### ensuring it all works with vscode

again, once we are happy with where the annotations render etc, ill feed this PR into my jsonschema scripts and try to figure how best to use the annotations for extensibility.

File: tools/protodoc/protodoc.py [hunk 10]
```diff
@@ -194,16 +220,47 @@ def FormatExtension(extension):
...
+  """Format extension metadata as RST.
+
+  Args:
+    extension_type: the name of the extension_type, e.g. com.acme.
+
+  Returns:
+    RST formatted extension type description.
+  """
+  IGNORED_EXTENSIONS = [
+    "envoy.tracers.xray", "envoy.filters.http.cache.simple_http_cache"]
```
Comment (phlax) [inline tools/protodoc/protodoc.py:248]: @htuch 

...also, there were 2 extensions that are in the `extension_db` but dont seem to use `envoy_cc_extension` func

ive added this to ignore them here for now, but not sure if this is the correct resolution

File: api/envoy/extensions/wasm/v3/wasm.proto [hunk 11]
```diff
@@ -59,6 +59,7 @@ message VmConfig {
   // (proxy_on_start). `google.protobuf.Struct` is serialized as JSON before
   // passing it to the plugin. `google.protobuf.BytesValue` and
   // `google.protobuf.StringValue` are passed directly without the wrapper.
+  // [#protodoc-extension: envoy.wasm.runtime]
```
Comment (phlax) [inline api/envoy/extensions/wasm/v3/wasm.proto:62]: the wasm plugins dont seem to have their own protos - i guess this is because the wasm runtime conf is passed as a  string through.

im wondering if there is any point to adding this here, the links it creates link to the runtime section immediately above and dont have additional config info

in terms of vscode/jsonschema, i guess it just needs to respect `Any`

File: tools/protodoc/protodoc.py [hunk 12]
```diff
@@ -194,16 +220,47 @@ def FormatExtension(extension):
     anchor = FormatAnchor('extension_' + extension)
     status = EXTENSION_STATUS_VALUES.get(extension_metadata['status'], '')
     security_posture = EXTENSION_SECURITY_POSTURES[extension_metadata['security_posture']]
+    extension_categories = "\n".join(
+      "  - :ref:`%s <extension_type_%s>`" % (ext, ext)
```
Comment (phlax) [inline tools/protodoc/protodoc.py:224]: test to see if f string works

File: tools/api_proto_plugin/annotations.py [hunk 13]
```diff
@@ -10,6 +10,9 @@
 # Page/section titles with special prefixes in the proto comments
 DOC_TITLE_ANNOTATION = 'protodoc-title'
 
+# Specify the type of extensions at extension points
+EXTENSION_TYPE_ANNOTATION = 'protodoc-extension'
```
Comment (phlax) [inline tools/api_proto_plugin/annotations.py:14]: the annotation for extension is just `extension` im wondering if this should be `extension-type` (ie no protodoc prefix)

File: docs/generate_extension_db.py [hunk 14]
```diff
@@ -37,18 +38,33 @@ def GetExtensionMetadata(target):
...
     raise ExtensionDbError(
         'Missing security posture for %s.  Please make sure the target is an envoy_cc_extension and security_posture is set'
         % target)
+  # evaluate tuples/lists
+  # wrap strings in a list
+  # use empty list for missing (this should not be necessary)
+  categories = (
+    '[]'
+    if IsMissing(categories)
+    else categories)
```
Comment (phlax) [inline docs/generate_extension_db.py:57]: try removing this ^^

File: docs/generate_extension_cats_db.py [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+def main():
+    extension_db = sys.argv[1]
+    extension_cat_db = sys.argv[2]
+
+    with open(extension_db) as f:
+        data = json.load(f)
+
+    categories = {}
+    for k, v in data.items():
+        print(">> %s" % k)
```
Comment (phlax) [inline docs/generate_extension_cats_db.py:16]: remove prints

File: docs/generate_extension_db.py [hunk 16]
```diff
@@ -37,18 +38,33 @@ def GetExtensionMetadata(target):
...
+  categories = (
+    ast.literal_eval(categories)
+    if ('[' in categories
+        or '(' in categories)
+    else [categories])
```
Comment (phlax) [inline docs/generate_extension_db.py:62]: this duck typing + eval is not ideal, but im struggling to think of a way of using the bazel query and allowing cats to be strings or arrays

File: tools/protodoc/protodoc.py [hunk 17]
```diff
@@ -194,16 +220,47 @@ def FormatExtension(extension):
...
+
+  Returns:
+    RST formatted extension type description.
+  """
+  IGNORED_EXTENSIONS = [
+    "envoy.tracers.xray", "envoy.filters.http.cache.simple_http_cache"]
+  try:
+    anchor = FormatAnchor('extension_type_' + extension_type)
+    extensions = "\n".join(
+      "  - :ref:`%s <extension_%s>`" % (ext, ext)
```
Comment (phlax) [inline tools/protodoc/protodoc.py:252]: this code is repeated, factor out

File: tools/protodoc/protodoc.py [hunk 18]
```diff
@@ -223,11 +280,16 @@ def FormatHeaderFromFile(style, source_code_info, proto_name):
   if annotations.EXTENSION_ANNOTATION in source_code_info.file_level_annotations:
     extension = source_code_info.file_level_annotations[annotations.EXTENSION_ANNOTATION]
     formatted_extension = FormatExtension(extension)
+  formatted_extension_type = ''
+  if annotations.EXTENSION_TYPE_ANNOTATION in source_code_info.file_level_annotations:
```
Comment (phlax) [inline tools/protodoc/protodoc.py:284]: not sure if this code is actually hit...

File: api/envoy/extensions/wasm/v3/wasm.proto (hunk 11 above)
Comment (htuch) [reply api/envoy/extensions/wasm/v3/wasm.proto:62]: @lizan how come these differ from the usual story that each extension point needs to be typed. The choice of runtime should have a typed config, right?

File: tools/api_proto_plugin/annotations.py (hunk 13 above)
Comment (htuch) [reply tools/api_proto_plugin/annotations.py:14]: +1

File: tools/protodoc/protodoc.py [hunk 19]
```diff
@@ -57,12 +57,31 @@
 This extension may be referenced by the qualified name *$extension*
 
 .. note::
+  EXTENSION_MARKER
```
Comment (phlax) [inline tools/protodoc/protodoc.py:60]: remove

File: tools/protodoc/protodoc.py [hunk 20]
```diff
@@ -57,12 +57,31 @@
...
+
+$categories
+
 """)
 
+
+EXTENSION_TYPE_TEMPLATE = string.Template("""$anchor
+
+.. tip::
+  EXTENSION_TYPE_MARKER
```
Comment (phlax) [inline tools/protodoc/protodoc.py:76]: remove

File: tools/protodoc/protodoc.py [hunk 21]
```diff
@@ -57,12 +57,31 @@
...
+
+  This extension type has the following known extensions
```
Comment (phlax) [inline tools/protodoc/protodoc.py:78]: im a bit concerned about nomenclature - ie extension, type, category etc it would be good to fix on a particular terminology

File: tools/protodoc/protodoc.py (hunk 10 above)
Comment (htuch) [reply tools/protodoc/protodoc.py:248]: Prefer to move them to `envoy_cc_extension` if possible.

**Review by htuch — COMMENTED** [review]
This is fantastic. I held back on picking on nits, but did a full design and extension anchor review, looks great.

File: tools/protodoc/protodoc.py (hunk 21 above)
Comment (htuch) [reply tools/protodoc/protodoc.py:78]: Extension is an extension, e.g. the HTTP buffer filter.

I think we should use category to refer to "HTTP filter". Drop the idea of "extension type" across the board.

File: tools/api_proto_plugin/annotations.py (hunk 13 above)
Comment (phlax) [reply tools/api_proto_plugin/annotations.py]: updated

File: tools/protodoc/protodoc.py (hunk 21 above)
Comment (phlax) [reply tools/protodoc/protodoc.py]: updated `type` -> `category` across PR

**Comment by phlax** [comment]
im struggling a bit with the formatting/linting here - its defo not python - and its forcing me to change legible code into long lines of very-hard-to-grok-code

at the same time its not even picking up basic linting like redundant imports

**Comment by htuch** [comment]
@phlax fixes welcome to Python linting/formatting, I don't think we are too set in stone here as long as we match Envoy style.

File: generated_api_shadow/BUILD [hunk 22]
```diff
@@ -145,7 +145,6 @@ proto_library(
         "//envoy/config/resource_monitor/injected_resource/v2alpha:pkg",
         "//envoy/config/retry/omit_canary_hosts/v2:pkg",
         "//envoy/config/retry/previous_hosts/v2:pkg",
-        "//envoy/config/retry/previous_hosts/v3:pkg",
```
Comment (phlax) [inline generated_api_shadow/BUILD:148]: @htuch not sure why this is necessary - seems to be unrelated to this PR as im seeing similar on others

**Comment by phlax** [comment]
> @phlax fixes welcome to Python linting/formatting, I don't think we are too set in stone here as long as we match Envoy style.

ok, ill raise a PR separately once this has gone through

File: tools/protodoc/protodoc.py (hunk 10 above)
Comment (phlax) [reply tools/protodoc/protodoc.py]: from further investigation it seems like what was missing wasnt the `envoy_cc_extension` entry after all - but the extension annotation in the relevant proto - ill add these and flag them...

File: api/envoy/config/trace/v3/xray.proto [hunk 23]
```diff
@@ -21,6 +21,7 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // [#protodoc-title: AWS X-Ray Tracer Configuration]
 // Configuration for AWS X-Ray tracer
 
+// [#extension: envoy.tracers.xray]
```
Comment (phlax) [inline api/envoy/config/trace/v3/xray.proto:24]: `envoy.tracers.xray` anno added here!

File: api/envoy/extensions/filters/http/cache/v3alpha/cache.proto [hunk 24]
```diff
@@ -18,8 +18,8 @@ option (udpa.annotations.file_status).work_in_progress = true;
 option (udpa.annotations.file_status).package_version_status = ACTIVE;
 
 // [#protodoc-title: HTTP Cache Filter]
-// [#extension: envoy.filters.http.cache]
 
+// [#extension: envoy.filters.http.cache]
```
Comment (phlax) [inline api/envoy/extensions/filters/http/cache/v3alpha/cache.proto:22]: seems like thiss one was just a ws typo on the annotation

Comment (phlax) [reply api/envoy/extensions/filters/http/cache/v3alpha/cache.proto]: ive had to append `.simple_http_cache` to this one - im guessing this is correct

File: docs/generate_extension_cat_db.py [hunk 25]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env python3
```
Comment (htuch) [inline docs/generate_extension_cat_db.py:1]: Can you add a top-level comment explaining what this does?

File: docs/generate_extension_cat_db.py [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+  extension_db = sys.argv[1]
+  extension_cat_db = sys.argv[2]
+
+  with open(extension_db) as f:
+    data = json.load(f)
+
+  categories = {}
+  for k, v in data.items():
+    for cat in v['categories']:
+      categories[cat] = categories.get(cat, [])
```
Comment (htuch) [inline docs/generate_extension_cat_db.py:17]: Suggest using a `defaultdict` here to simplify.

File: docs/generate_extension_db.py [hunk 27]
```diff
@@ -53,18 +54,25 @@ def GetExtensionMetadata(target):
...
+  rout = r.stdout.decode('utf-8').strip().split(' ')
+  security_posture, status, undocumented = rout[:3]
+  categories = ' '.join(rout[3:])
   if IsMissing(security_posture):
     raise ExtensionDbError(
         'Missing security posture for %s.  Please make sure the target is an envoy_cc_extension and security_posture is set'
         % target)
+  # evaluate tuples/lists
+  # wrap strings in a list
+  categories = (ast.literal_eval(categories) if
```
Comment (htuch) [inline docs/generate_extension_db.py:69]: It's gross, I like it!

File: tools/protodoc/protodoc.py [hunk 28]
```diff
@@ -61,6 +61,19 @@
 
   $security_posture
 
+  This extension extends the following extension category:
+
+$categories
```
Comment (htuch) [inline tools/protodoc/protodoc.py:66]: As discussed offline, suggest moving outside the note.

File: tools/protodoc/protodoc.py [hunk 29]
```diff
@@ -194,16 +214,43 @@ def FormatExtension(extension):
...
+    extension_category: the name of the extension_category, e.g. com.acme.
+
+  Returns:
+    RST formatted extension category description.
+  """
+  try:
+    extensions = EXTENSION_CATEGORIES[extension_category]
+  except KeyError as e:
+    sys.stderr.write(f"\n\nUnable to find extension category:  {extension_category}\n\n")
+    exit(1)  # Raising the error buries the above message in tracebacks.
```
Comment (htuch) [inline tools/protodoc/protodoc.py:248]: Suggest turning the KeyError into a ProtodocError and then throwing.

**Review by htuch — CHANGES_REQUESTED** [review]
Looks awesome. Some minor comments/nit and this can ship. Very exciting to see it come together.
/wait

File: tools/protodoc/protodoc.py [hunk 30]
```diff
@@ -92,6 +105,15 @@
...
 }
 
+EXTENSION_DB = json.loads(pathlib.Path(os.getenv('EXTENSION_DB_PATH')).read_text())
+
+# create an index of extension categories from extension db
+EXTENSION_CATEGORIES = {}
+for _k, _v in EXTENSION_DB.items():
+  for _cat in _v['categories']:
+    EXTENSION_CATEGORIES.setdefault(_cat, [])
+    EXTENSION_CATEGORIES[_cat].append(_k)
```
Comment (phlax) [inline tools/protodoc/protodoc.py:115]: @htuch i removed the script and shifted the code here

this simplifies/reduces the code and also parses the extension db only once, which should improve the efficiency of protodoc

i have used setdefault as `defaultdict` adds some complexity and would prevent `KeyErrors` from being raised below

File: docs/generate_extension_cat_db.py (hunk 26 above)
Comment (phlax) [reply docs/generate_extension_cat_db.py]: i shifted this code elsewhere, and slightly reformulated...

File: docs/generate_extension_cat_db.py (hunk 25 above)
Comment (phlax) [reply docs/generate_extension_cat_db.py]: comment added in new location of code

File: tools/protodoc/protodoc.py (hunk 29 above)
Comment (phlax) [reply tools/protodoc/protodoc.py:253]: the problem is that we need the `exit` - i can make protodocerror inherit `SysExit` tho - should i ?

Comment (phlax) [reply tools/protodoc/protodoc.py:253]: hmm - maybe it *will* exit anyway - ill test...

Comment (phlax) [reply tools/protodoc/protodoc.py]: cool - raising `ProtodocError` actually gives better err/tb - updated

File: tools/protodoc/protodoc.py (hunk 28 above)
Comment (phlax) [reply tools/protodoc/protodoc.py:71]: ive shifted this into a `tip` also - partly because the equivalence with the `tip` in the categories is helpful in terms of navigation - but also because i think it needs some kind of highlighting of "this is where this can go"

**Comment by phlax** [comment]
@htuch nits addressed (hopefully)

File: tools/protodoc/protodoc.py [hunk 31]
```diff
@@ -189,21 +221,51 @@ def FormatExtension(extension):
...
+
+  Returns:
+    RST formatted extension category description.
+  """
+  try:
+    extensions = EXTENSION_CATEGORIES[extension_category]
+  except KeyError as e:
+    raise ProtodocError(f"\n\nUnable to find extension category:  {extension_category}\n\n")
+  anchor = FormatAnchor('extension_category_' + extension_category)
+  extensions = FormatExtensionList(sorted(extensions))
```
Comment (phlax) [inline tools/protodoc/protodoc.py:265]: move `sorted` into the extension list func

**Review by htuch — APPROVED** [review]
LGTM, thanks! So cool.

**Comment by mattklein123** [comment]
@phlax this is absolutely amazing. Well done. For future work, I'm wondering if we should do the following:
1) Also document the type_url?
2) Potentially auto generate YAML snippets of the name/type_url? I think there might still be confusion on how to actually do the configuration.

We should track these improvements in other issues? WDYT?

**Comment by phlax** [comment]
> We should track these improvements in other issues? WDYT?

i think it sounds like a good idea - altho from where i am working right now, im not entirely clear if we are talking about documenting end use (ie `@type`) or dev (ie `type_url` + `value`)

so, there are a couple of existing tickets in relation to this.

i opened this one:

https://github.com/envoyproxy/envoy/issues/13531

mostly born from frustration of finding configuration examples using the older `config` and finding difficult to find information on how to upgrade, or where the definitions for the newer configs could be found

for the first part of the problem - upgrading - i guess its a lessening issue as time  passes - perhaps we should just close it focus forward

the second part of it is where to get accurate info on extensions etc, which i think is the existing ticket that you had opened previously:

https://github.com/envoyproxy/envoy/issues/13167

i think that one has a partial fix with this PR - lets use it to track further improvements on the extension <> category docs
